src/iot/commands/createCert.ts

```diff
@@ -0,0 +1,133 @@
+    window = Window.vscode(),
+    commands = Commands.vscode()
+): Promise<void> {
+    getLogger().debug('CreateCertificate called for %O', node)
+
+    const isConfirmed = await showConfirmationMessage(
+        {
+            prompt: localize('AWS.iot.createCert.prompt', 'Create a new X.509 certificate and RSA key pair?'),
+            confirm: localize('AWS.iot.createCert.confirm', 'Confirm'),
+            cancel: localizedText.cancel,
+        },
+        window
+    )
+    if (!isConfirmed) {
+        getLogger().info('CreateCertificate canceled')
+        return
+    }
+
+    const folderLocation = await promptForSaveLocation(window)
+    if (!folderLocation) {
+        getLogger().info('CreateCertificate canceled: No folder selected')
+        return
+    }
+    const certPath = `${folderLocation.fsPath}-certificate.pem.crt`
+    const privateKeyPath = `${folderLocation.fsPath}-private.pem.key`
+    const publicKeyPath = `${folderLocation.fsPath}-public.pem.key`
+
+    const certExists = await fileExists(certPath)
+    const privateKeyExists = await fileExists(privateKeyPath)
+    const publicKeyExists = await fileExists(publicKeyPath)
+
+    if (certExists) {
+        getLogger().error('Certificate path {0} already exists', certPath)
+        window.showErrorMessage(
+            localize('AWS.iot.createCert.error', 'Failed to create certificate. Path {0} already exists.', certPath)
+        )
+        return undefined
+    }
+    if (privateKeyExists) {
+        getLogger().error('Key path {0} already exists', privateKeyPath)
+        window.showErrorMessage(
+            localize(
+                'AWS.iot.createCert.error',
+                'Failed to create certificate. Path {0} already exists.',
+                privateKeyPath
+            )
+        )
+        return undefined
+    }
+    if (publicKeyExists) {
+        getLogger().error('Key path {0} already exists', publicKeyPath)
+        window.showErrorMessage(
+            localize(
+                'AWS.iot.createCert.error',
+                'Failed to create certificate. Path {0} already exists.',
+                publicKeyPath
+            )
+        )
+        return undefined
+    }
+
+    try {
+        const certificate = await node.iot.createCertificateAndKeys({
+            setAsActive: false,
+        })
+        const certId = certificate.certificateId
+        const certPem = certificate.certificatePem
+        const privateKey = certificate.keyPair?.PrivateKey
+        const publicKey = certificate.keyPair?.PublicKey
+
+        if (!certPem || !privateKey || !publicKey) {
+            getLogger().error('Could not download certificate. Certificate is missing either the PEM or key pair.')
+            showViewLogsMessage(localize('AWS.iot.createCert.error', 'Failed to create certificate'), window)
+            return undefined
+        }
+
+        //Save resources
+        await fs.writeFile(certPath, certPem, { encoding: PEM_FILE_ENCODING, mode: MODE_RW_R_R })
```

Comment:
I did not write a unit test for this file, as I was unsure how to mock out this filesystem write, or if this should be written to a temporary directory. Is there an example of a test for a function like this?

Reply:
So far we've been almost always making real FS writes, though it may be time to start adding a bit of abstraction to the file system.

> Key pairs cannot be re-downloaded from AWS, so they are lost forever if they're not saved, and so the user would need to delete the certificate from their account

In this case we may want to prompt the user with the key pair string to manually save it (maybe a button to save to clipboard for each one?). Read-only ephemeral documents may work too. Will need to look into this more for what the best practice ought to be.

Reply:
For now I added a block that deletes the certificate from AWS if the key pair was not saved.